data-structures/unixfsv2.md

```diff
@@ -0,0 +1,68 @@
+# UnixFSv2 Schema
+
+This doc contains the File and Directory schemas for UnixFSv2.
+
+This schema depends on some Advanced Layouts and types defined
+in other schemas:
+
+* [Data](https://github.com/ipld/specs/pull/211)
+* [Hashmap](https://github.com/ipld/specs/blob/master/data-structures/hashmap.md)
+
+```sh
+type DirData union {
+  | EntryMap "map"
+  | Hamt "hamt"
+} representation keyed
+
+type EntryMap {String:EntryUnion}
+
+type EntryUnion union {
+  | File "file"
+  | &File "fileLink"
+  | Directory "dir"
+  | &Directory "dirLink"
+} representation keyed
+
+# WIP: Hashmap does not yet have an advanced layout definition.
```

Comment:
The intention of the current schema spec is that you can just do this:

```
advanced HAMT
type Hamt {String:FileUnion} representation advanced HAMT
```

Although I believe `s/FileUnion/EntryUnion` in the current iteration.


What else do we need to proceed with baking this in?

Reply:
i had assumed that the library you publish would already have `advanced HAMT` in it because you need something to attach the implementation of the advanced layout to before handing it off to a consumer. i would then consume that w/ `type DirHamt {String:FileUnion} representation advanced HAMT`

Reply:
Yeah, that would work too, this is a really fuzzy area (as partly demonstrated by #208). I've been imagining `advanced HAMT` as a signalling mechanism that "this thing exists somewhere and will need to be hooked up" and that comes from the thinking that also included `advanced HAMT { implementation "something" }`. It bothers me that we don't have a good answer to where these things sit and what's required where.

So I suppose to be maximally minimal you could omit `advanced HAMT`.

When I was doing up the Filecoin spec docs (have since deleted that work so I can't show it), I found that having `advanced HAMT; advanced Sharray` were useful documentation tools, saying "you'll encounter these things as you work through the rest of the docs, when you do, come back here and you'll understand that we're pointing to some adjacent space that has an implementation of these things. Like a named break-point in the schema that can be coupled with its own documentation. Not quite as helpful here because you only have one instance and could put the docs in that one place to explain what that is.